modules/data.atmosphere/R/download.NOAA_GEFS_downscale.R

```diff
@@ -0,0 +1,341 @@
+##' }
+##' 
+##' @author Katie Zarada - modified code from Luke Dramko and Laura Puckett
+##' 
+##' 
+
+
+download.NOAA_GEFS_downscale <- function(outfolder, lat.in, lon.in, sitename, start_date = Sys.time(), end_date = (as.POSIXct(start_date, tz="UTC") + lubridate::days(16)),
+                                         overwrite = FALSE, verbose = FALSE, ...) {
+  
+  start_date <- as.POSIXct(start_date, tz = "UTC")
+  end_date <- as.POSIXct(end_date, tz = "UTC")
+  
+  #It takes about 2 hours for NOAA GEFS weather data to be posted.  Therefore, if a request is made within that 2 hour window,
+  #we instead want to adjust the start time to the previous forecast, which is the most recent one avaliable.  (For example, if
+  #there's a request at 7:00 a.m., the data isn't up yet, so the function grabs the data at midnight instead.)
+  if (abs(as.numeric(Sys.time() - start_date, units="hours")) <= 2) {
+    start_date = start_date - lubridate::hours(2)
+    end_date = end_date - lubridate::hours(2)
+  }
+  
+  #Date/time error checking - Checks to see if the start date is before the end date
+  if (start_date > end_date) {
+    PEcAn.logger::logger.severe("Invalid dates: end date occurs before start date")
+  } else if (as.numeric(end_date - start_date, units="hours") < 6) { #Done separately to produce a more helpful error message.
+    PEcAn.logger::logger.severe("Times not far enough appart for a forecast to fall between them.  Forecasts occur every six hours; make sure start 
+                                and end dates are at least 6 hours appart.")
+  }
+  
+  #Set the end forecast date (default is the full 16 days)
+  if (end_date > start_date + lubridate::days(16)) {
+    end_date = start_date + lubridate::days(16)
+  }
+  
+  #Round the starting date/time down to the previous block of 6 hours.  Adjust the time frame to match.
+  forecast_hour = (lubridate::hour(start_date) %/% 6) * 6 #Integer division by 6 followed by re-multiplication acts like a "floor function" for multiples of 6
+  increments = as.integer(as.numeric(end_date - start_date, units = "hours") / 6) #Calculating the number of forecasts between start and end dates.
+  increments = increments + ((lubridate::hour(end_date) - lubridate::hour(start_date)) %/% 6) #These calculations are required to use the rnoaa package.
+  
+  end_hour = sprintf("%04d", ((forecast_hour + (increments * 6)) %% 24) * 100)  #Calculating the starting hour as a string, which is required type to access the 
+  #data via the rnoaa package
+  forecast_hour = sprintf("%04d", forecast_hour * 100)  #Having the end date as a string is useful later, too.
+  
+  #Recreate the adjusted start and end dates.
+  start_date = as.POSIXct(paste0(lubridate::year(start_date), "-", lubridate::month(start_date), "-", lubridate::day(start_date), " ", 
+                                 substring(forecast_hour, 1,2), ":00:00"), tz="UTC")
```

Comment:
FYI you can also use the base R function `ISOdatetime` to construct date-times from individual components:

```r
ISOdatetime(lubridate::year(start_date), lubridate::month(start_date), lubridate::day(start_date), lubridate::hour(start_date), 0, 0, tz = "UTC")
```

See `?ISOdatetime` for details.